### ISSUE TITLE
``N[F[Pi],prec,Method->method]`` does evaluates to ``F[N[Pi,prec,Method->method]]``

### ISSUE THREAD

**Issue by mmatera** [Mathics3/mathics-core#137]
The use of the `Method` option for numeric evaluation breaks the behavior or `N`, that should propagate to its leaves.
For example, ``N[F[Pi],10]`` does evaluates to ``F[N[Pi,10]]`` produces the (``TraceEvaluation``):
```
    Evaluating: System`N[Global`F[System`Pi], 10]
      Evaluating: System`N
      Evaluating: Global`F[System`Pi]
        Evaluating: Global`F
        Evaluating: System`Pi
           [...]
      Evaluating: System`N[10]
         [...]
      -> 10.0
      Evaluating: System`N[Global`F, 10]
        [...]
      -> Global`F
      Evaluating: System`N[System`Pi, 10]
        [...]
      -> 3.141592654
    -> Global`F[3.141592654]
      Evaluating: Global`F
F[3.141592654]
```
But, ``N[F[Pi],10, Method->"numpy"]`` does evaluates to ``F[N[Pi,10]]``  produces (``TraceEvaluation``):

```
```
    Evaluating: System`N[Global`F[System`Pi], 10, System`Rule[System`Method, "numpy"]]
      Evaluating: System`N
      Evaluating: Global`F[System`Pi]
        Evaluating: Global`F
        Evaluating: System`Pi
      Evaluating: System`Rule[System`Method, "numpy"]
        Evaluating: System`Rule
        Evaluating: System`Method
N[F[Pi], 10, Method -> numpy]
```


```

**Comment by mmatera** [comment]
There are different possibilities to handle ``N[expr, prec, Method->mthd]``:

* Just add a rule to discard the option, so 
`N[F[Pi],10, Method->"numpy"]`->`N[F[Pi],10]`->`F[N[Pi,10]]`->`F[N[Pi,10,Method->Automatic]`
* N[expr, prec, Method->mthd] could be evaluated to  N[expr, prec], but setting an internal variable in `Evaluation`, that can be inspected for `N[const,prec]` when `method` is supported.
* Evaluate  `N[F[leaf1, ...], prec, Method->mthd]` to  ``N[F, prec, Method->mthd][N[leaf1, prec, Method->mthd],...]

I guess the second option would be optimal.

**Comment by rocky** [comment]
I spent the last 10 minutes going over this trying to figure out what you are talking about. For now I have to give up. 

I bet if you spent 10 minutes you would be able to find an even smaller example. LIke the example that is so small that if you remove something it no longer has that behavior.  To paraphrase Einstein: make as simple as possible but no simpler. 

The is a very useful programming and debugging skill.


> But, `N[F[Pi],10, Method->"numpy"]` does evaluates to `F[N[Pi,10]]` produces (TraceEvaluation):
>  
```
Evaluating: System`N[Global`F[System`Pi], 10, System`Rule[System`Method, "Automatic"]]
                                                                          ^^^^^ Automatic? 
      Evaluating: System`N
```

The other suggestion is to leave this until after we go over how evaluation works in general and come up with a description of the process.

**Comment by mmatera** [comment]
@rocky, there is no smaller example. Maybe what I should have to do is to invest more time explaining the reason for mentioning this issue. 

In WL, rules of the form `N[expr_,prec_]` associated with a symbol are special: are the so called ``NValues``.  In Mathics, `NValues` are stored in a different list that the `upvalues`, in a way that if you try to evaluate
``F[x]`` rules of the form ``N[expr, prec]`` are not considered (but  are considered rules of the form ``g[x]``, for example).

`NValues` rules are just applied in the (builitin) apply method of ``N``. When an expression of the form
``N[F[x]]`` is evaluated, first it is converted to ``N[F[x], $MachinePrecision]`` and afterward, the built-in method is called.

This method (among other things) look for the `NValues` of `F` and try to apply the rules. If fails, tries to evaluate ``N[F,$MachinePrecision][N[x, $MachinePrecision], $MachinePrecision]``

in a way that, in the minimum example  ``N[F[Pi]]``  the following chain of replacements takes place:

``N[F[Pi]]`` ->``N[F[Pi], $MachinePrecision]``->``N[F, $MachinePrecision][  N[Pi, $MachinePrecision]]]` -> 
``F[3.1415926626`]``

Now, in WMA, ``N`` does not accept options, and Mathics neither until we implemented them in `mathics.builtin.numbers.constants`. In the current implementation, something like `N[x_, OptionsPattern[]]` is stored as an `NValue`, while `N[x_,prec_, OptionsPattern[]]` is stored as an `UpValue`. This can break the logic behind `mathics.builtin.numeric.N.apply_with_prec`: the second argument in an `NValue` rule should be a number, not an option.

If now the ``Method->`` option is passed to ``N`` (without precision specification)
``N[F[Pi], Method->"numpy"]``

the current implementation tries to use `Method->"numpy"` as a precision specification, and the evaluation fails with the error
"N::precbd: Requested precision Method -> numpy is not a machine-sized real number."
returning  ``N[F[Pi], Method->"numpy"]``. 

On the other hand, if you pass the precision parameter `N[F[Pi], 10, Method->"numpy"]`, the default rules for ``N``  do not match with the expression, and the expression stays unevaluated.

**Comment by mmatera** [comment]
Regarding this

> 
> ```
> Evaluating: System`N[Global`F[System`Pi], 10, System`Rule[System`Method, "Automatic"]]
>                                                                           ^^^^^ Automatic? 
>       Evaluating: System`N
> ```

It was a mistake: the actual  trace in the current master branch would be:
```
    Evaluating: System`N[Global`F[System`Pi], 10, System`Rule[System`Method, "numpy"]]
      Evaluating: System`N
      Evaluating: Global`F[System`Pi]
        Evaluating: Global`F
        Evaluating: System`Pi
      Evaluating: System`Rule[System`Method, "numpy"]
        Evaluating: System`Rule
        Evaluating: System`Method
N[F[Pi], 10, Method -> numpy]
```

the trace I put in the description comes from a modified branch, where I tried to allow `mathics.builtin.numeric.N.apply_with_prec` to accept an optional value.

**Comment by mmatera** [comment]
> 
> The other suggestion is to leave this until after we go over how evaluation works in general and come up with a concise description of the process. With that we'll be able to fit this behavior into the general process.

One nice thing about looking at the ``N.apply_with_prec`` method instead the general ``evaluation`` method is that numerical evaluation is "simpler": there are fewer attributes to have into account, fewer types of rules, and any improvement here would have an impact on the performance of  Graphics, all without touching almost anything on the ``mathics.core`` side.

**Comment by rocky** [comment]
> @rocky, there is no smaller example. Maybe what I should have to do is to invest more time explaining the reason for mentioning this issue.

Yes, thanks for doing this. I appreciate the time and effort here. 

> 
> In WL, rules of the form `N[expr_,prec_]` associated with a symbol are special: are the so called `NValues`. In Mathics, `NValues` are stored in a different list tha[n] the `upvalues`, in a way that if you try to evaluate `F[x]` rules of the form `N[expr, prec]` are not considered (but are considered rules of the form `g[x]`, for example).

Okay. If this "different list" isn't called NValues, we might consider naming it NValues. 

> 
> `NValues` rules are just applied in the (builitin) apply method of `N`. When an expression of the form `N[F[x]]` is evaluated, first it is converted to `N[F[x], $MachinePrecision]` and afterward, the built-in method is called.

Okay. This seems just like having a transformation rule added to the `N[]` function.  Nothing out of the ordinary here. 
> 
> This method (among other things) look for the `NValues` of `F` and try to apply the rules. If fails, tries to evaluate `N[F,$MachinePrecision][N[x, $MachinePrecision], $MachinePrecision]`
> 
> in a way that, in the minimum example `N[F[Pi]]` the following chain of replacements takes place:
> 
> `N[F[Pi]]` ->`N[F[Pi], $MachinePrecision]`->`` N[F, $MachinePrecision][ N[Pi, $MachinePrecision]]]` -> ``F[3.1415926626`]``
> 

Got it. This makes perfect sense. The `N[]` function has a rule that it applies to each argument. I believe we handle that kind of thing already, right? 


> Now, in WMA, `N` does not accept options, and Mathics neither until we implemented them in `mathics.builtin.numbers.constants`. In the current implementation, something like `N[x_, OptionsPattern[]]` is stored as an `NValue`, while `N[x_,prec_, OptionsPattern[]]` is stored as an `UpValue`. This can break the logic behind `mathics.builtin.numeric.N.apply_with_prec`: the second argument in an `NValue` rule should be a number, not an option.
> 

Ok. So we (or rather probably I), added this "Method" option not understanding all of these nuances. 

> If now the `Method->` option is passed to `N` (without precision specification) `N[F[Pi], Method->"numpy"]`
> 
> the current implementation tries to use `Method->"numpy"` as a precision specification, and the evaluation fails with the error "N::precbd: Requested precision Method -> numpy is not a machine-sized real number." returning `N[F[Pi], Method->"numpy"]`.
> 
> On the other hand, if you pass the precision parameter `N[F[Pi], 10, Method->"numpy"]`, the default rules for `N` do not match with the expression, and the expression stays unevaluated.

Ok. So to me the remedy is simple. Remove the "Method" option, since it wasn't done in a way that was well thought out to be consistent with WL. 

--- 

Let us not go down a rabbit hole of adding more gadgets and complications right now.

In my opinion, what I see needs to happen in a parallel branch is to _simplify_ and modularize thing is that is not the case.  

Specifically:

* write down, find, or cull together a technical description of how WL evaluation works that a person and program
* come up with an implementation that matches this, including adding the rules with built in `N[]` and so on. If we can use the existing implementation or maybe with small changes and refactors, that's great. But I would strip out initially stuff like caching, not because I don't like caching, I love it, but I don't want there to be a chance that it is correcting for behavior that would otherwise be wrong were it not there. This is very important.

What would be different this time over the last time &mdash; and this is very important &mdash; is that we'd have a modular, well documented, logically simple, and explainable code. Also &mdash; and this is very important &mdash; we would start having _tests_ for the subsystems and ways to interact with it in smaller pieces. Sort of what we did in pytests and the mathics.session object. 

After all of this we are in a position to benchmark parts of it. And add back in and go over the things that speed things up.  And we are in a position to benchmark how good each of these is because we will have more mastery over the subsystems and can benchmark subsystems in isolation as well as embedded in the full execution. 

> One nice thing about looking at the N.apply_with_prec method instead the general evaluation method is that numerical evaluation is "simpler": there are fewer attributes to have into account, fewer types of rules, and any improvement here would have an impact on the performance of Graphics, all without touching almost anything on the mathics.core side.

Do me a favor and please forget about performance. What I seen time and time again is spending time over stuff like whether `hasattr()` is faster than  `is` or the order of `elif` tests when we are missing the fact that we are going through the entire evaluation loop an additional time.  We are an order of magnitude slower than we should be. And I think the reason is that there are high-level blunders going on. 

We need an approach to addressing that. We need more master at a higher-level of the code

**Comment by mmatera** [comment]
Fixed by #138
